### PR TITLE
Chore: Fix github workflow and let only member runs mimir-build-image

### PR DESCRIPTION
**Pull request by ying-jeanne**
#### What this PR does
This PR arises from the decision to disable builds for contributors outside of the Mimir organization. Every contributor's pipeline execution must now receive approval from a maintainer. This change was questioned by discussions in another PR.

What's the rationale behind this approach? It's rooted in an issue opened within the dependabot-core repository (https://github.com/dependabot/dependabot-core/issues/3253). I have implemented two workflows: one for membership checking, executed on `pull_request_target`, which ensures that it operates exclusively within the context of the target branch. This means that no modifications in the commit branch are considered until they are merged into the target main/release branch. Consequently, contributors cannot bypass the membership check. This behaviour has been tested on my personal repository (https://github.com/ying-jeanne/mimir/pull/107).

Additionally, if the PR creator is not a part of the Grafana organization, the CI will fail and no build starts, as demonstrated in my personal test (https://github.com/ying-jeanne/mimir/actions/runs/6659976149/job/18100191154?pr=85). However, re-running the job using the 'ying-jeanne' account resulted in success (https://github.com/ying-jeanne/mimir/actions/runs/6659976149?pr=85).

Note the second workflow is not part of CI success, It's worth noting that the `build mimir-build-image` workflow is adding comment to indicate a new build is ongoing and prompts them to wait. If this approach proves insufficient, we are open to further enhancements.

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- na Tests updated
- na Documentation added
- na `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by ying-jeanne** [comment]
Another option is to trigger the build image on merge to target branch, then raise another PR for review. Link to the PR that trigger the change.

File: .github/workflows/member_pr.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            id: teamAffiliation
+            run: |
+                # Encode the triggering actor's username for use in the GitHub API URL. Otherwise the user like renovate[bot] has encoding error
+                ENCODED_USERNAME=$(printf %s "${{ github.triggering_actor }}" | xxd -p -c 256 | tr -d '\n' | sed 's/\(..\)/%\1/g')
+                
+                # Define the GitHub API URL for checking team membership.
+                API_URL="https://api.github.com/orgs/grafana/teams/mimir-maintainers/members/$ENCODED_USERNAME"
+                
+                # Send a GET request to the GitHub API with proper authentication.
+                response=$(curl -s -o /dev/null -w "%{http_code}" -H "Authorization: token ${{ secrets.GH_BOT_ACCESS_TOKEN }}" $API_URL)
```
Comment (aknuds1) [inline .github/workflows/member_pr.yml:30]: Could you instead use `gh api orgs/grafana/teams/mimir-maintainers/members/${{ github.triggering_actor }}` here? Ought to be much simpler than going via `curl`.

File: .github/workflows/member_pr.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            run: | 
+                gh pr comment $PR_NUMBER --body "Changing the Dockerfile for mimir-build-image requires approval from a mimir-maintainer. Please reach out to one of the mimir-maintainers to review your changes before proceeding."
+            env:
+                PR_NUMBER: ${{ github.event.pull_request.number }}
+                GITHUB_TOKEN: ${{secrets.GH_BOT_ACCESS_TOKEN}}
+
+          - name: Stop workflow if user is no member
+            if: ${{ steps.teamAffiliation.outputs.team_membership != 'true' }}
+            run: |
+              echo "You have no rights to trigger this job."
```
Comment (aknuds1) [inline .github/workflows/member_pr.yml:58]: "You lack permission" sounds a little less aggressive :)
```suggestion
              echo "You lack permission to trigger this job."
```

**Review by aknuds1 — COMMENTED** [review]
I think I understand your scheme, and it makes sense to me. Please see comments.

File: .github/workflows/member_pr.yml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                
+                # Define the GitHub API URL for checking team membership.
+                API_URL="https://api.github.com/orgs/grafana/teams/mimir-maintainers/members/$ENCODED_USERNAME"
+                
+                # Send a GET request to the GitHub API with proper authentication.
+                response=$(curl -s -o /dev/null -w "%{http_code}" -H "Authorization: token ${{ secrets.GH_BOT_ACCESS_TOKEN }}" $API_URL)
+                
+                # Check the response and update the result in the GitHub output.
+                if [ -z "$response" ]; then
+                  echo "team_membership=error" >> $GITHUB_OUTPUT
```
[the remark this answers is not in the document]
Comment (ying-jeanne) [reply .github/workflows/member_pr.yml:34]: it gives error when response is empty or not integer on line just after

**Comment by dimitarvdimitrov** [comment]
is this PR still needed? since we merged https://github.com/grafana/mimir/pull/6071 and agreed that pushing an image from a fork is not a risk?

**Comment by ying-jeanne** [comment]
> is this PR still needed? since we merged #6071 and agreed that pushing an image from a fork is not a risk?

It is no longer needed. close it up. Thanks for the reminder.

**Pull request closed without merging**